python/pyspark/serializers.py

```diff
@@ -224,7 +225,13 @@ def _create_batch(series):
     # If a nullable integer series has been promoted to floating point with NaNs, need to cast
     # NOTE: this is not necessary with Arrow >= 0.7
     def cast_series(s, t):
-        if t is None or s.dtype == t.to_pandas_dtype():
+        if type(t) == pa.TimestampType:
+            # NOTE: convert to 'us' with astype here, unit ignored in `from_pandas` see ARROW-1680
+            return _series_convert_timestamps_internal(s).values.astype('datetime64[us]')
+        elif t == pa.date32():
+            # TODO: ValueError: Cannot cast DatetimeIndex to dtype datetime64[D]
```

Comment:
I came across an issue when using `pandas_udf` that returns a `DateType`.  Dates are input to the udf as Series with dtype `datetime64[ns]` and trying to use this for `pa.Array.from_pandas` with `type=pa.date32()` fails with an error.  I am also unable to call `series.dt.values.astype('datetime64[D]')` which results in an error.  Without specifying the type, pyarrow will read the values as a timestamp.  I filed https://issues.apache.org/jira/browse/ARROW-1718 to look into this, aside from this being fixed in a new version any ideas for a workaround @ueshin and @HyukjinKwon ?
cc @wesm

Reply:
How about:

```diff
@@ -229,14 +229,20 @@ def _create_batch(series):
             # NOTE: convert to 'us' with astype here, unit ignored in `from_pandas` see ARROW-1680
             return _series_convert_timestamps_internal(s).values.astype('datetime64[us]')
         elif t == pa.date32():
-            # TODO: ValueError: Cannot cast DatetimeIndex to dtype datetime64[D]
-            return s.dt.values.astype('datetime64[D]')
+            return s.dt.date
         elif t is None or s.dtype == t.to_pandas_dtype():
             return s
         else:
             return s.fillna(0).astype(t.to_pandas_dtype(), copy=False)

-    arrs = [pa.Array.from_pandas(cast_series(s, t), mask=s.isnull(), type=t) for s, t in series]
+    def create_array(s, t):
+        casted = cast_series(s, t)
+        if casted.dtype == 'object':
+            return pa.Array.from_pandas(casted, type=t)
+        else:
+            return pa.Array.from_pandas(casted, mask=s.isnull(), type=t)
+
+    arrs = [create_array(s, t) for s, t in series]
     return pa.RecordBatch.from_arrays(arrs, ["_%d" % i for i in xrange(len(arrs))])
```

Reply:
It seems we need to remove `mask=s.isnull()` when dtype is `object` because Arrow raises an error saying `ArrowNotImplementedError: NotImplemented: mask not supported in object conversions yet`.

Reply:
There's a small number of places where this is not yet supported, but we should be able to fix them in 0.8.0 (~next 2 weeks): https://issues.apache.org/jira/browse/ARROW-1721

Reply:
Thanks @ueshin and @wesm, this seems to work.  I will add a note with the related JIRA to look into it later